Icon Autoslicing

LemonInTheDark opened the pull request:
## About The Pull Request

Ok so you know all the dmis we have that are made to work with the smoothing system? carpets, walls, etc.

The proper way to edit those is to convert them into a png with 5 "states' it in (one for 0 connections, one for horizontal, one for vertical, one for all cardinals and one for all directions) and then modify THAT, then run it through [the cutter tool.](https://github.com/tgstation/icon-cutter)

But none ever does that, because we explain it fucking nowhere. So instead, let's keep all those "base" files in the repo, alongside the configs they work with, and "cut" the pngs into dmis as a part of the build process.

I wrote a guide for how to interact with this system as a spriter, you can find it [HERE](https://github.com/LemonInTheDark/tgstation/blob/slice-the-sky/icons/Cutter.md).

[Adds a icon cutter build task](https://github.com/tgstation/tgstation/commit/52143d2e96498de92421d516e0dd3f23936f88d8)

This relies on action ninja's hypnagogic (find more [here](https://github.com/actioninja/hypnagogic)), a rust based icon cutter.
It operates inline with the file structure, searching the codebase for templates and resource files and compiling them down to dmis.

It can do way more then just bitmask stuff, but that is what we are using it for rn.

Hope is to prevent for eternity the "I'm just gonna edit each of these 255 icon states that's how this carpet was made right?" meme, and allow more expansive use of smoothing in future

[Adds a lint that ensures config files work right](https://github.com/tgstation/tgstation/commit/21eeab9cf831c5fdac5a9b366478a9dab285c20c)

Checks to ensure they have a paired png and dmi, and also avoids issues with uncompiled changes by double checking that nothing happens before/after a cutter run

[Pulls all non smoothed states out of structures into bespoke dmis](https://github.com/tgstation/tgstation/commit/a730e0cb47fc0a622fe265bccc296cec8d3a8fea)

This is required because the cutter cannot output named icon states, only the actual cut icon

[Does something similar to walls](https://github.com/tgstation/tgstation/commit/40780e9481103c8ee9e16538d1c2d0cdc124eeb9)

Moves reinforced walls decon stuff from their icon to a var on the type and a set of states in the reinforced_states dmi

Moves falsewalls into their own dmi, this involved some changes to gamecode to ensure falsewalls knew which dmi to use and what key.
Makes falsewalls display as such in editor rather then just walls

Moves smoothrock's gibonite overlays into their own file for similar reasons

[Same thing different day (Floors)](https://github.com/tgstation/tgstation/commit/9a3da3b69705278f39af109ac5ce86d27c2479a1)

Pulls bespoke floor icon states into their own file, splits up neon carpets into multiple files to make cutting possible

[Actually adds the cut templates and their matching png files](https://github.com/tgstation/tgstation/commit/1bd8920dc90d1ee1b934b6dadc39f2331854f5fa)

Not much to report here, outside of I changed the prefix for bamboo walls to bamboo_wall so it works with false_walls

## Why It's Good For The Game

![image](https://github.com/tgstation/tgstation/assets/58055496/7c3ac7fb-873c-481b-8667-082e39432876)

None should have to manually edit cut dmis. Ever.
Also this makes adding a new smoothed thing trivial, don't even need to know what tool you're using to do it. V good v good.
Sets us up nicely for wallening's well, wall of sprites.

Some structural decisions, we are essentially committing build artifacts here. That's the best way of handling it because otherwise mappers could need to run build.bat before opening a map, and that is stupid!

## Changelog
:cl:
refactor: (Almost) all smoothed icons can now be edited in their pre cut forms
/:cl:

LemonInTheDark committed:
Adds a icon cutter build task

This relies on action ninja's hypnagogic, a rust based icon cutter.
It operates inline with the file structure, searching the codebase for
templates and resource files and compiling them down to dmis.

It can do way more then just bitmask stuff, but that is what we are
using it for rn.

Hope is to prevent for eternity the "I'm just gonna edit each of these
255 icon states that's how this carpet was made right?" meme, and allow
    more expansive use of smoothing in future

LemonInTheDark committed:
Adds a lint that ensures config files work right

Checks to ensure they have a paired png and dmi, and also avoids issues
with uncompiled changes by double checking that nothing happens
before/after a cutter run

LemonInTheDark committed:
Pulls all non smoothed states out of structures into bespoke dmis

This is required because the cutter cannot output named icon states,
only the actual cut icon

LemonInTheDark committed:
Does something similar to walls

Moves reinforced walls decon stuff from their icon to a var on the type
and a set of states in the reinforced_states dmi

Moves falsewalls into their own dmi, this involved some changes to
gamecode to ensure falsewalls knew which dmi to use and what key.
Makes falsewalls display as such in editor rather then just walls

Moves smoothrock's gibonite overlays into their own file for similar
reasons

LemonInTheDark committed:
Same thing different day (Floors)

Pulls bespoke floor icon states into their own file, splits up neon
carpets into multiple files to make cutting possible

LemonInTheDark committed:
whoops shitty name + neon carpets impl

LemonInTheDark committed:
fixes up some templates

LemonInTheDark committed:
Actually adds the cut templates and their matching png files

Not much to report here, outside of I changed the prefix for bamboo
walls to bamboo_wall so it works with false_walls

LemonInTheDark committed:
whoops missed the pngs

LemonInTheDark committed:
ensures check.py actually works

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into slice-the-sky

LemonInTheDark committed:
touchups

LemonInTheDark committed:
I hate linux

LemonInTheDark committed:
chmod?

LemonInTheDark committed:
bumps all runners to 22.04

LemonInTheDark commented:
Showing a bunch of diffs because the cutter does that for some reason. It ought to be a one time thing, if it wasn't the linter would throw an error.

LemonInTheDark commented:
https://github.com/tgstation/tgstation/pull/79659/checks?check_run_id=18585633101

Some larger icons (snow) seem to be drifting. need to look inot that

MrStonedOne commented on tools/icon_cutter/check.py:
```diff
@@ -0,0 +1,57 @@
+
+BUF_SIZE = 65536  # lets read stuff in 64kb chunks!
+
+def hash_file(path):
```
don't hash the file, hash the actual contents of the png image data block. you'll get drifting over time from implementation differences wrt to pixel encoding (png supports 8 different ways of encoding the rgba values including palettes) and compression (the image data block is gzip compressed and different implementations or versions of the same implementations or just different oses or even env vars can cause it to compress at different gzip compression levels.).

use some python lib to convert the png to an array of 4byte values and hash that.

LemonInTheDark replied:
excellent idea thank you.
what I'm doing now is using pillow to pull metadata, then pull the image data as a list of tuples and hash it as a string.

not sure if there's a better way of doing about that though

LemonInTheDark replied:
worked out a way that's impervious to png dimension changes

LemonInTheDark replied:
handled, we pull metadata and compare/output diffs (kinda ugly tho idk), then read the png as an image, split it into a flat array that doesn't care about dimensions, and then compares

LemonInTheDark committed:
Instead of hashing dmis as raw files, hashes their metadata and pixels.

Hopefully this avoids false positives, should make shittesting this pr
easier too

LemonInTheDark committed:
include pngs in that too

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into slice-the-sky

LemonInTheDark committed:
that shouldn't be there

LemonInTheDark committed:
back_to_basics

LemonInTheDark committed:
Adds more complete error handling to the checker

LemonInTheDark committed:
Properly implements meat wall, fixes up bamboo some, clears the wall dmis of all the animations and cruft

LemonInTheDark committed:
does something similar to structures

LemonInTheDark committed:
updates clockwork png to the right colors

LemonInTheDark committed:
Fixes up the icerock png, I think? time green made it weird as a joke

LemonInTheDark committed:
Properly names all the carpet states

LemonInTheDark committed:
Final cut with all current changes, sendin er

LemonInTheDark commented:
Alright. I've fixed up all the diffs I introduced accidentially, and improved the lint.
There's like, one thing that timegreen made by I think hand that I need confirmation on, then we should be good pending review

LemonInTheDark commented:
Of note:
The cutter increases file size pretty signficantly. This is because the crate it uses, (dmi, maintained by the spacestation13 org), uses the image crate from rust which does not provide an easy api for using any png compression. I've talked to zewaka a maintainer about this, gonna look into it real briefly, might just whackjob it.

LemonInTheDark commented:
Oh also in the process of going from initial dmi -> template png (I used a cutter format) I ended up changing a few pixels by like 1 rgb value. I think it's a result of how exporting from dmis works, I can go through each instance and fix them up if you'd like but I don't really consider it worth my time.

LemonInTheDark committed:
numpy requirement

LemonInTheDark committed:
hopefully? fixes tests

LemonInTheDark committed:
tests my beloved

LemonInTheDark committed:
whoop\

LemonInTheDark committed:
are you happy now?

LemonInTheDark committed:
clears dev images

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into slice-the-sky

LemonInTheDark commented:
Waiting on https://github.com/actioninja/hypnagogic/pull/3

LemonInTheDark committed:
Reduces sizes with new cutter version

LemonInTheDark committed:
bump cutter version

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into slice-the-sky

LemonInTheDark commented:
we ball

Time-Green commented:
Looks good! 
![image](https://github.com/tgstation/tgstation/assets/7501474/04c2c976-a5de-424c-864a-f0cdc1dc7ed2)
THere's a bit of a weird icon glitch there, but also that used to be the entire map so that's a pretty big win

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into slice-the-sky

LemonInTheDark committed:
markdown file

san7890 commented on code/game/turfs/closed/minerals.dm:
```diff
@@ -659,7 +659,7 @@
 /turf/closed/mineral/gibtonite
 	mineralAmt = 1
-	icon_state = "rock_Gibtonite_inactive"
+	MAP_SWITCH(, icon_state = "rock_Gibtonite_inactive")
```
```suggestion
	MAP_SWITCH(null, icon_state = "rock_Gibtonite_inactive")
```

this is the pattern you use elsewhere so we should keep it consistent. IMO we should just make a fully new macro instead of doing this but whatever

LemonInTheDark replied:
I am switching between setting a var and not setting it. null would just runtime.
W 515 I could use :: here off parent_type, tho that might break downstreams

san7890 commented on code/game/turfs/closed/minerals.dm:
```diff
@@ -764,7 +764,7 @@
 
 /turf/closed/mineral/gibtonite/ice
-	icon_state = "icerock_Gibtonite_inactive"
+	MAP_SWITCH(, icon_state = "icerock_Gibtonite_inactive")
```
```suggestion
	MAP_SWITCH(null, icon_state = "icerock_Gibtonite_inactive")
```

LemonInTheDark replied:
see above

LemonInTheDark committed:
better docs

LemonInTheDark committed:
proper sizing

LemonInTheDark committed:
better

LemonInTheDark committed:
more better

LemonInTheDark committed:
output example

LemonInTheDark committed:
Removes unneeded config details

LemonInTheDark committed:
declutter

LemonInTheDark committed:
better comments

LemonInTheDark committed:
output_name stuff

LemonInTheDark committed:
clarity on file names

LemonInTheDark committed:
standard -> cardinal

LemonInTheDark committed:
syntax

LemonInTheDark committed:
hh

LemonInTheDark committed:
whoop

optimumtact commented:
This broke the build process for literally everyone because hypnagogic has a hard dependency on glibc that isn't present on ubuntu or windows.

optimumtact commented:
I'm not even sure why hypnagogic is in the build path, i thought this PR was about committing the build results and not part of the juke build standard build.

LemonInTheDark commented:
spoke on discord but:
It's in the build path because I want as easy a pattern for spriters as possible. (I don't trust them to run a tool basically, and I want this to be a smooth experience even if they do know to). It is a build input so it seems sane to use juke.

The trouble is juke does its "should we rebuild" based off file modification stuff (as I understand). So if our outputs are "older" then our inputs, we rerun. Git doesn't carry this between builds, so you get a recompile even if you don't touch anything.
I could fix this by using like a file that holds a hash of inputs or something but I didn't. (didn't occur to me)

We commit the artifacts so mappers and such don't need to run build.bat to get a working build. 

The dependency issue is something action is working to fix, will update to a new release as soon as that gets done.

LemonInTheDark commented:
forrr the record we should no longer be dependent on glibc, I have changed the cutter's releases to target more portable compilers and have it set to statically compile a bit more
